File: Inprotech.Web/Configuration/Rules/Checklists/ChecklistMaintenanceService.cs
using System;
using Inprotech.Infrastructure.Security;
using Inprotech.Infrastructure.Validations;
using InprotechKaizen.Model;
using InprotechKaizen.Model.Components.Configuration;
using InprotechKaizen.Model.Persistence;
using InprotechKaizen.Model.Rules;

namespace Inprotech.Web.Configuration.Rules.Checklists
{
    public interface IChecklistMaintenanceService
    {
        dynamic CreateChecklistCriteria(ChecklistSaveModel formData);
    }

    public class ChecklistMaintenanceService : IChecklistMaintenanceService
    {
        readonly IDbContext _dbContext;
        readonly ILastInternalCodeGenerator _lastInternalCodeGenerator;
        readonly ICriteriaMaintenanceValidator _maintenanceValidator;
        readonly ITaskSecurityProvider _taskSecurityProvider;

        public ChecklistMaintenanceService(IDbContext dbContext,
                                           ILastInternalCodeGenerator lastInternalCodeGenerator,
                                           ICriteriaMaintenanceValidator maintenanceValidator,
                                           ITaskSecurityProvider taskSecurityProvider)
        {
            _dbContext = dbContext;
            _lastInternalCodeGenerator = lastInternalCodeGenerator;
            _maintenanceValidator = maintenanceValidator;
            _taskSecurityProvider = taskSecurityProvider;
        }

        public dynamic CreateChecklistCriteria(ChecklistSaveModel formData)
        {
            ValidationError validation;

            if ((validation = _maintenanceValidator.ValidateCriteriaName(formData.CriteriaName)) != null)
            {
                return _maintenanceValidator.Error(validation);
            }

            if (formData.IsProtected && !_taskSecurityProvider.HasAccessTo(ApplicationTask.MaintainCpassRules, ApplicationTaskAccessLevel.Create))
            {
                throw new Exception("does not have permission to create protected rules");
            }

            var criteria = new Criteria
            {
                IsProtected = formData.IsProtected,
                Description = formData.CriteriaName,
                RuleInUse = formData.InUse ? 1 : 0,
                OfficeId = formData.Office,
                CaseTypeId = formData.CaseType,
                CountryId = formData.Jurisdiction,
                PropertyTypeId = formData.PropertyType,
                CaseCategoryId = formData.CaseCategory,
                SubTypeId = formData.SubType,
                BasisId = formData.Basis,
                Profile = formData.Profile,
                ChecklistType = formData.Checklist,
                PurposeCode = CriteriaPurposeCodes.CheckList
            }.WithUnknownToDefault();
            if (formData.IsLocalClient == null)
            {
                criteria.LocalClientFlag = null;
            }
            else
            {
                criteria.LocalClientFlag = formData.IsLocalClient.Value ? 1 : 0;
            }

            if ((validation = _maintenanceValidator.ValidateDuplicateCriteria(criteria, true)) != null)
            {
                return _maintenanceValidator.Error(validation);
            }

            criteria.Id = _lastInternalCodeGenerator.GenerateLastInternalCode(KnownInternalCodeTable.Criteria);

            _dbContext.Set<Criteria>().Add(criteria);
            _dbContext.SaveChanges();

            return new
            {
                Status = true,
                CriteriaId = criteria.Id
            };
        }
    }
}

File: condor/client/ng/configuration/rules/checklists/maintenance/checklist-maintenance.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';

@Injectable({
    providedIn: 'root'
})
export class ChecklistMaintenanceService {

    baseUrl = 'api/configuration/rules/checklist-configuration/';
    constructor(private readonly http: HttpClient) { }

    createChecklist(data: any): Observable<any> {

        return this.http.put(`${this.baseUrl}add`, this.buildCriteria(data));
    }

    private readonly buildCriteria = (criteria: any) => {
        if (criteria == null) {
            return undefined;
        }

        return {
            criteriaName: criteria.criteriaName,
            caseType: this.getKey(criteria, 'caseType', 'code'),
            caseCategory: this.getKey(criteria, 'caseCategory', 'code'),
            caseProgram: this.getKey(criteria, 'program', 'key'),
            jurisdiction: this.getKey(criteria, 'jurisdiction', 'code'),
            propertyType: this.getKey(criteria, 'propertyType', 'code'),
            subType: this.getKey(criteria, 'subType', 'code'),
            basis: this.getKey(criteria, 'basis', 'code'),
            office: this.getKey(criteria, 'office', 'key'),
            profile: this.getKey(criteria, 'profile', 'code'),
            checklist: this.getKey(criteria, 'checklist', 'key'),
            isProtected: criteria.isProtected,
            inUse: criteria.isInUse
        };
    };

    private readonly getKey = (searchCriteria: any, propertyName: string, key: string) => {
        return searchCriteria[propertyName] && searchCriteria[propertyName][key];
    };

}
